// @sap/html5-repo-mock/lib/services/html5-apps-repo/discovery.js
const fs = require('fs').promises;
const path = require('path');
const console = require('../../utils/logger.js');
const pom = require('./pom.js');
const APP_NAME = /^[a-zA-Z0-9_ ]+$/;
const APP_VERSION = /^(\d+\.){2}(\d+)$/;
const HOST_BY_PORT = 'http://localhost:3001/AppStudio/api/getHostByPort?port=';
let appFrontBaseURL;

const clearCache = () => {
    appFrontBaseURL = undefined;
};

const getAppFrontBaseURL = async ({arURL, isAppStudio}) => {
    if (!appFrontBaseURL) {
        // If running in SAP Business Application Studio environment
        appFrontBaseURL = new URL(arURL);
        if (isAppStudio) {
            appFrontBaseURL = new URL(
                await fetch(HOST_BY_PORT + appFrontBaseURL.port)
                    .then(res => res.json())
                    .then(data => data.result), 
                arURL
            );
        } else {
            appFrontBaseURL.hostname = `port${appFrontBaseURL.port}.` + appFrontBaseURL.hostname;
        }
    }
    return appFrontBaseURL;
};

const getURLWithSubdomainPrefix = 
    (baseURL, prefix) => baseURL.protocol + '//' + prefix + '-' + baseURL.host + baseURL.pathname;

const listFiles = async (basePath) => {
    const dirContents = await fs.readdir(basePath, { withFileTypes: true });
    let filePaths = [];
    for(let i=0; i<dirContents.length; i++) {
        const name = path.join(basePath, dirContents[i].name);
        if(dirContents[i].isFile()) {
            filePaths.push(name);
        } else if (dirContents[i].isDirectory()) {
            filePaths = filePaths.concat(await listFiles(name));
        }
    }
    return filePaths;
};

const discover = async ({dir, lookupDirs, arURL, isAppStudio}) => {
    const baseDirs = (await fs.readdir(dir[0], { withFileTypes: true })).reduce((acc, cur) => {
        if(cur.isDirectory()) {
            acc.push(path.join(dir[0], cur.name));
        }
        return acc;
    }, dir.slice(0));
    await getAppFrontBaseURL({arURL, isAppStudio});
    return (await Promise.all(baseDirs.map(async baseDir => {
        try {
            const xsappPath = path.join(baseDir, 'xs-app.json');
            await fs.access(xsappPath);
            console.debug(`The '${baseDir}' contains xs-app.json and is a candidate for being HTML5 application`);
            const xsapp = JSON.parse(await fs.readFile(xsappPath, 'utf8'));
            const isAppFront = xsapp?.routes?.some?.(route => route.service === 'app-front');
            const interpolate = await pom(path.join(baseDir, 'pom.xml'));
            for (let lookupDir of lookupDirs) {
                const manifestPath = path.join(baseDir, lookupDir, 'manifest.json');
                try {
                    const manifest = JSON.parse(await fs.readFile(manifestPath, 'utf8'));
                    const appId = interpolate(manifest['sap.app'].id);
                    const appName = appId.replace(/[.-]/g, '');
                    const appVersion = interpolate(
                        manifest['sap.app'].applicationVersion.version
                    ).replace('-SNAPSHOT', '');
                    const sapCloudService = manifest['sap.cloud'] && manifest['sap.cloud'].service || '';
                    const isPublic = !!(manifest['sap.cloud'] && manifest['sap.cloud'].public);
                    if (!appName.match(APP_NAME)) {
                        return console.error(
                            `Error: ${manifestPath} contains invalid sap.app/id = "${manifest['sap.app'].id}". `+
                            'Only [a-zA-Z0-9.-_ ] are allowed (e.g. "my.app_Name1").\n'
                        );
                    }
                    if (!appVersion.match(APP_VERSION)) {
                        return console.error(
                            `Error: ${manifestPath} contains invalid sap.app/applicationVersion/version `+
                            `= "${appVersion}". Only three not negative numbers separated by dots are allowed `+
                            '(e.g. "1.0.0").\n'
                        );
                    }
                    const resourcesDir = path.join(baseDir, lookupDir);
                    console.debug(
                        `The ${appId}@${appVersion} application found. ` +
                        `Static resources will be served from '${resourcesDir}' directory`
                    );
                    return {
                        applicationId: appId,
                        applicationName: appName,
                        applicationVersion: appVersion,
                        sapCloudService: sapCloudService,
                        'public': isPublic,
                        __conf: baseDir,
                        __dir: resourcesDir,
                        __af: isAppFront,
                        __url: isAppFront ? 
                            getURLWithSubdomainPrefix(
                                appFrontBaseURL, 
                                `app-${appName}-v${appVersion.replaceAll('.','_')}`
                            ):
                            `${arURL}${appName}-${appVersion}/`
                    };
                } catch(err) { /* Do nothing */ }
            }
        } catch(err) { /* Do nothing */ }
        return undefined;
    }))).filter(m => !!m);
};

module.exports.listFiles = listFiles;
module.exports.discover = discover;
module.exports.clearCache = clearCache;
