// @sap/html5-repo-mock/lib/services/app-front/index.js
const console = require('../../utils/logger.js');

const appFrontMiddleware = (req, res, next) => {
    if (req.url.startsWith('/app-front') && req.headers['x-forwarded-path']) {
        console.debug(
            `Rewriting '${req.url}' to '${req.headers['x-forwarded-path']}'`+
            `(x-correlationid: ${req.headers['x-correlationid']})`
        );
        req.url = req.headers['x-forwarded-path'];
        res.setHeader('Access-Control-Allow-Origin', req.headers['origin'] ?? '*');
        res.setHeader('Access-Control-Allow-Headers', '*');
        res.setHeader('Access-Control-Allow-Methods', '*');
        res.setHeader('Access-Control-Allow-Credentials', 'true');
    }
    next();
};

module.exports = async () => {
    const { add } = require('../../utils/vcap.js');
    add({
        service: 'app-front',
        plan: 'mock',
        tags: ['app-front'],
        sapCloudService: 'app-front',
        grantType: 'client_credentials',
        path: '/app-front'
    });
    return appFrontMiddleware;
};
